File: 30. Substring with Concatenation of All Words Hard/solution.py
from collections import defaultdict

def find_substring(s: str, words: list[str]) -> list[int]:
    if not words[0]:
        return []
    
    word_length = len(words[0])
    word_count = len(words)
    word_frequency = defaultdict(int)
    
    for word in words:
        word_frequency[word] += 1
    
    result = []
    
    for i in range(word_length):
        left = i
        right = i
        current_count = 0
        current_frequency = defaultdict(int)
        
        while right + word_length <= len(s):
            word = s[right:right + word_length]
            right += word_length
            
            if word in word_frequency:
                current_frequency[word] += 1
                
                current_count += 1
                
                while current_frequency[word] > word_frequency[word]:
                    left_word = s[left:left + word_length]
                    current_frequency[left_word] -= 1
                    left += word_length
                    current_count -= 1
                
                if current_count == word_count:
                    result.append(left)
            else:
                current_frequency.clear()
                current_count = 0
                left = right
    
    return result


print(find_substring("barfoothefoobarman", ["foo","bar"]))
print(find_substring("wordgoodgoodgoodbestword", ["word","good","best","word"]))
print(find_substring("barfoofoobarthefoobarman", ["bar","foo","the"]))
print(find_substring("wordgoodgoodgoodbestword", ["word","good","best","good"]))

File: 30. Substring with Concatenation of All Words Hard/solution.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(findSubstring("barfoothefoobarman", []string{"foo", "bar"}))
	fmt.Println(findSubstring("wordgoodgoodgoodbestword", []string{"word", "good", "best", "word"}))
	fmt.Println(findSubstring("barfoofoobarthefoobarman", []string{"bar", "foo", "the"}))
	fmt.Println(findSubstring("wordgoodgoodgoodbestword", []string{"word", "good", "best", "good"}))
}

func findSubstring(s string, words []string) []int {
	if len(words) == 0 {
		return []int{}
	}

	wordLength := len(words[0])
	wordCount := len(words)
	wordFrequency := map[string]int{}

	for _, word := range words {
		wordFrequency[word]++
	}

	result := []int{}

	for i := 0; i < wordLength; i++ {
		left := i
		right := i
		currentCount := 0
		currentFrequency := map[string]int{}

		for right+wordLength <= len(s) {
			word := s[right : right+wordLength]
			right += wordLength

			if wordFrequency[word] != 0 {
				currentFrequency[word]++
				currentCount++

				for currentFrequency[word] > wordFrequency[word] {
					leftWord := s[left : left+wordLength]
					currentFrequency[leftWord]--
					left += wordLength
					currentCount--
				}

				if currentCount == wordCount {
					result = append(result, left)
				}
			} else {
				currentFrequency = map[string]int{}
				currentCount = 0
				left = right
			}
		}
	}

	return result
}
